lib/Runtime/Library/JavascriptBigInt.cpp

```diff
@@ -397,4 +414,183 @@ namespace Js
     }
 
+    // pbi1 += pbi2 assume pbi1 > pbi2
+    void JavascriptBigInt::AddAbsolute(JavascriptBigInt * pbi1, JavascriptBigInt * pbi2)
+    {
+        digit_t carryDigit = 0;
+        digit_t *pDigit1 = pbi1->m_digits;
+        digit_t *pDigit2 = pbi2->m_digits;
+        digit_t i = 0;
+
+        for (; i < pbi2->m_length; i++)
+        {
+            digit_t tempCarryDigit = 0;
+            if (pDigit2[i] > 0)
+            {
+                pDigit1[i] = JavascriptBigInt::AddDigit(pDigit1[i], pDigit2[i], &tempCarryDigit);
+            }
+            if (carryDigit > 0)
+            {
+                pDigit1[i] = JavascriptBigInt::AddDigit(pDigit1[i], carryDigit, &tempCarryDigit);
+            }
+            carryDigit = tempCarryDigit;
+        }
+
+        for (; i < pbi1->m_length; i++)
+        {
+            digit_t tempCarryDigit = 0;
+            if (carryDigit > 0)
+            {
+                pDigit1[i] = JavascriptBigInt::AddDigit(pDigit1[i], carryDigit, &tempCarryDigit);
+            }
+            carryDigit = tempCarryDigit;
+        }
+
+        if (0 < carryDigit) // length increase by 1
+        {
+            if (pbi1->m_length >= pbi1->m_maxLength && !pbi1->Resize(pbi1->m_length + 1))
+            {
+                return;
+            }
+            pbi1->m_digits[pbi1->m_length++] = carryDigit;
+        }
+    }
+
+    // pbi1 -= pbi2 assume pbi1 > pbi2
+    void JavascriptBigInt::SubAbsolute(JavascriptBigInt * pbi1, JavascriptBigInt * pbi2)
+    {
+        digit_t carryDigit = 0;
+        digit_t *pDigit1 = pbi1->m_digits;
+        digit_t *pDigit2 = pbi2->m_digits;
+        digit_t i = 0;
+
+        for (; i < pbi2->m_length; i++)
+        {
+            digit_t tempCarryDigit = 0;
+            if (pDigit2[i] > 0)
+            {
+                pDigit1[i] = JavascriptBigInt::SubDigit(pDigit1[i], pDigit2[i], &tempCarryDigit);
+            }
+            if (carryDigit > 0)
+            {
+                pDigit1[i] = JavascriptBigInt::SubDigit(pDigit1[i], carryDigit, &tempCarryDigit);
+            }
+            carryDigit = tempCarryDigit;
+        }
+
+        for (; i < pbi1->m_length; i++)
+        {
+            digit_t tempCarryDigit = 0;
+            if (carryDigit > 0)
+            {
+                pDigit1[i] = JavascriptBigInt::SubDigit(pDigit1[i], carryDigit, &tempCarryDigit);
+            }
+            carryDigit = tempCarryDigit;
+            if (carryDigit == 0)
+            {
+                break;
+            }
+        }
+
+        Assert(carryDigit == 0);
+        // adjust length
+        while ((pbi1->m_length>0) && (pbi1->m_digits[pbi1->m_length-1] == 0))
+        {
+            pbi1->m_length--;
+        }
+        Assert(pbi1->m_length > 0);
+    }
+
+    JavascriptBigInt * JavascriptBigInt::Sub(JavascriptBigInt * pbi1, JavascriptBigInt * pbi2)
+    {
+        if (JavascriptBigInt::IsZero(pbi1))
+        {
+            pbi2->m_isNegative = !pbi2->m_isNegative;
```

Comment:
pbi2 shouldn't be modifiable at this point. We need to clone it before negating.

Reply:
The `const` keyword would probably be useful in preventing this sort of bug. These top-level methods could take `const JavascriptBigInt*`, and then to get a copy that the compiler will let you modify, you'd have to call `JavascriptBigInt::New`.

---
In reply to: [227860626](https://github.com/Microsoft/ChakraCore/pull/5792#discussion_r227860626) [](ancestors = 227860626)

Reply:
Also probably worth including some cases in assign_by_value.js that verify that the second arg to addition and subtraction doesn't get modified.

---
In reply to: [227861565](https://github.com/Microsoft/ChakraCore/pull/5792#discussion_r227861565) [](ancestors = 227861565,227860626)